criu/util.c

```diff
@@ -1583,4 +1583,38 @@ char *get_legacy_iptables_bin(bool ipv6)
 	iptables_present[ipv6] = 1;
 
 	return iptables_bin[ipv6];
-}
\ No newline at end of file
+}
+
+ssize_t read_all(int fd, void *buf, size_t size)
+{
+	ssize_t n = 0;
+	while (n < size) {
+		ssize_t ret = read(fd, buf, size - n);
+		if (ret == -1) {
+			if (errno == EINTR)
+				continue;
+			return ret;
+		}
+		if (ret == 0)
+			return n;
+		n += ret;
+		buf = (char *)buf + ret;
+	}
+	return n;
+}
+
+ssize_t write_all(int fd, const void *buf, size_t size)
+{
+	ssize_t n = 0;
+	while (n < size) {
+		ssize_t ret = write(fd, buf, size - n);
+		if (ret == -1) {
+			if (errno == EINTR)
```

Comment:
In addition to `EINTR`, should we also check for `errno == EAGAIN`?

Reply:
So what happens when we've already made some progress, and then we get EAGAIN.

Should we return the number of bytes already written? in which case, this is not that better than the original write()

Should we return a EAGAIN? But then we've already written some content, and the caller will need to know about it. So that's bad. Can't do that.

In other words, if we can get EAGAIN, the caller is just better off using regular read()/write(). Which means that the read_all()/write_all() APIs are for blocking fds only.
In this case, if we get EAGAIN maybe it would be appropriate to return EINVAL? and add a comment on these functions that these APIs are for blocking fds only.
What do you think?

Reply:
>In this case, if we get EAGAIN maybe it would be appropriate to return EINVAL? and add a comment on these functions that these APIs are for blocking fds only.

Sounds good. read_all() and write_all() are generic functions and it would be good to keep them simple.

Reply:
I've updated the PR